libvast/builtins/formats/pcap.cpp

```diff
@@ -0,0 +1,510 @@
+        }
+        co_yield builder.finish();
+      }
+    };
+    return make(ctrl, std::move(input), !!args_.emit_file_header);
+  }
+
+  friend auto inspect(auto& f, pcap_parser& x) -> bool {
+    return f.object(x)
+      .pretty_name("pcap_parser")
+      .fields(f.field("args", x.args_));
+  }
+
+private:
+  parser_args args_;
+};
+
+struct printer_args {
+  // template <class Inspector>
+  // friend auto inspect(Inspector& f, printer_args& x) -> bool {
+  //   return f.object(x).pretty_name("printer_args");
+  // }
+};
+
+class pcap_printer final : public plugin_printer {
+public:
+  pcap_printer() = default;
+
+  explicit pcap_printer(printer_args args) : args_{std::move(args)} {
+  }
+
+  auto name() const -> std::string override {
+    return "pcap";
+  }
+
+  static auto pack(const file_header& header) -> chunk_ptr {
+    const auto* ptr = reinterpret_cast<const std::byte*>(&header);
+    auto bytes = std::span<const std::byte>{ptr, sizeof(file_header)};
+    return chunk::copy(bytes);
+  }
+
+  static auto make_file_header(const table_slice& slice)
+    -> std::optional<file_header> {
+    if (slice.schema().name() != "pcap.file_header" || slice.rows() == 0)
+      return std::nullopt;
+    auto result = file_header{};
+    const auto& input_record = caf::get<record_type>(slice.schema());
+    auto array = to_record_batch(slice)->ToStructArray().ValueOrDie();
+    auto xs = values(input_record, *array);
+    auto begin = xs.begin();
+    if (begin == xs.end() || *begin == std::nullopt)
+      return std::nullopt;
+    for (const auto& [key, value] : **begin) {
+      if (key == "magic_number")
+        result.magic_number
+          = detail::narrow_cast<uint32_t>(caf::get<uint64_t>(value));
+      if (key == "major_version")
+        result.major_version
+          = detail::narrow_cast<uint16_t>(caf::get<uint64_t>(value));
+      if (key == "minor_version")
+        result.minor_version
+          = detail::narrow_cast<uint16_t>(caf::get<uint64_t>(value));
+      if (key == "reserved1")
+        result.reserved1
+          = detail::narrow_cast<uint32_t>(caf::get<uint64_t>(value));
+      if (key == "reserved2")
+        result.reserved2
+          = detail::narrow_cast<uint32_t>(caf::get<uint64_t>(value));
+      if (key == "snaplen")
+        result.snaplen
+          = detail::narrow_cast<uint32_t>(caf::get<uint64_t>(value));
+      if (key == "linktype")
+        result.snaplen
+          = detail::narrow_cast<uint32_t>(caf::get<uint64_t>(value));
+    }
+    return result;
+  }
+
+  static auto
+  make_file_header(uint16_t linktype, uint32_t snaplen = maximum_snaplen)
+    -> std::optional<file_header> {
+    return file_header{
+      .magic_number = magic_number_2,
+      .major_version = 2,
+      .minor_version = 4,
+      .reserved1 = 0,
+      .reserved2 = 0,
+      .snaplen = snaplen,
+      .linktype = linktype,
+    };
+  }
+
+  auto instantiate(type input_schema, operator_control_plane& ctrl) const
```

Comment:
@tenzir/pipelines-squad I'm struggling with this error:

```
error: !! logic_error: 'print "pcap", pcap_printer(anonymous())' does not support heterogeneous outputs; cannot initialize for 'pcap' after 'pcap.packet'
 = note: from `print "pcap", pcap_printer(anonymous())`
```

My hunch is that I cannot dispatch in `instantiate` like I do. Can you advice how I can handle two different schema types properly?

Reply:
Your printer has `allows_joining() == false`. If the saver has `is_joining() == true` (e.g. `to file ...`), then we know that all output schemas go to the same stream, interleaved. But since your printer claims that it does not support this joining/interleaving, we can only give it events for a single schema. The error message above happens when there are events of multiple schemas.

Reply:
Makes sense. I enabled joining. What are the guarantees on concurrent execution? Can I share state between printers from the same call to `instantiate`? Say I return a different generator based on the schema name, and I'd like them to share state, is that allowed? (That's what I'm doing now.) Or should I rather always return one generator that does the schema-based dispatching internally?

Reply:
This is now obsolete. I found and fixed the issues.